### PR TITLE
Several Improvements

### PR DESCRIPTION
**Pull request by lpaulsen93**
This pull-request brings the following improvements:

- Bugfixes for #98 and #95
- Column names now may collide, they only need to be unique per plugin
- Now a plugin can add more than one column (parameter ```$column``` is passed to functions ```th()``` and ```td ()```
- A plugin can specify a CSS class per ```<th>``` or ```<td>``` element (parameter ```$class``` is passed to functions ```th()``` and ```td ()```

I also like to implement two new columns for the task list (=pagelist) of the task plugin. One is a text based priority column and the other one an image based priority column. So I need this PR to be checked/merged. Once to be able to add more than one column by the task plugin and another time for better styling of a single cell in the task/page list.

Changes are backwards compatible.

### CODE REVIEW

**Comment by LarsGit223** [comment]
Here is the PR of the task plugin: https://github.com/dokufreaks/plugin-task/pull/39.

File: helper.php [hunk 1]
```diff
@@ -420,10 +536,16 @@ function _diffCell($id) {
...
     /**
      * Plugins - respective plugins must be installed!
+     * @see addColumn for more information
      */
     function _pluginCell($plug, $col, $id) {
-        if (!isset($this->page[$col])) $this->page[$col] = $this->$plug->td($id);
-        return $this->_printCell($col, $this->page[$col]);
+        // At this point we do not need to differ between built-in and other plugins
+        $class = '';
+        if (!isset($this->page['plugins'][$plug][$col])) $this->page['plugins'][$plug][$col] = $this->$plug->td($id, $col, $class);
```
Comment (Klap-in) [inline helper.php]: Please use brackets

**Comment by Klap-in** [comment]
Hi!
Maybe I'm asking too fast, because I didn't read all the code, but:
What is that 'built-in' plugin stuff?

Is that td() and th() really needed?

My feeling is that the code of this plugin is already complex to understand, so adding these options aren't making it easier... So far I haven't suggestions to make the code (structure) simpler, but imho it is worth to thing about how things could be done simpler..

**Comment by LarsGit223** [comment]
Hi Klap-in,

no problem, just go ahead asking:
> What is that 'built-in' plugin stuff?

This is only needed to differ between plugins which are fixed built into the pagelist plugin ('discussion', 'linkback', 'tag', 'pageimage') and the one's that were dynamically added by a call to ```addColumn```. All plugins are listed in ```plugins```. The built-in plugins use the array ```header``` and ```column```. Unfortunately these are public so I could not extend or change them in a safe way. So I added ```pluginheaders``` and ```plugincolumns``` which are only used by the "dynamic" plugins added with ```addColumn```.

> Is that td() and th() really needed

Yes. That is actually the nice/extendible part of the plugin (also I would prefer to use an interface for that). For example have a look at the task plugin. It is using the normal header/columns for which the heading and the content is generated by the pagelist plugin itself (e.g. columns "page", "date", "user"...). Then the task plugin is adding own columns e.g. "status" and "priority". As the pagelist does not know hot to get the values for this foreign columns owned by other plugins, it is calling the ```td``` and ```th``` functions of that plugins. If you wouldn't have this mechanism then the pagelist would not be extendable at all.

I think it could get simpler if you would throw out the "built-in" plugins and make the plugin handle all plugins in the same way (also the pagelist plugin could add itself using ```addColumns()```). But I think that is not possible without breaking backwards compatibility. That's why I wrote the mail. I also have no idea how to implement this changes in a more beautiful way AND keep backwards compatibility.

So IMHO this is probably the last new feature that should be put in the existing code.

File: helper.php (hunk 1 above)
Comment (LarsGit223) [reply helper.php]: Added missing brackets to all if-statements.

**Comment by Klap-in** [comment]
Maybe it is for future use nicer to simplify things. And, therefore, introduce an simpler set of functions, which replace the other on long term?

**Comment by Klap-in** [comment]
I see no strange things. So if you like to merge it, it is fine for me.

**Comment by Klap-in** [comment]
Superseded by #130. However, in that PR the extending is done in a bit simplified manner.

**Comment by Klap-in** [comment]
Thanks for the PR, based on these ideas #130 and #139 are made.

**Pull request closed without merging**